src/Illuminate/Container/Container.php

```diff
@@ -1070,6 +1082,24 @@ protected function fireCallbackArray($object, array $callbacks)
         }
     }
 
+    /**
+     * Check if object or a generalisation is pending in the resolve stack.
+     *
+     * @param   object    $object
+     *
+     * @return  bool
+     */
+    protected function pendingInResolveStack($object)
+    {
+        foreach ($this->resolveStack as $resolving) {
```

Comment:
Not sure we have to resolve the last element on each loop. I would make sense to extract it to a variable before.
Or even slicing it.

Reply:
This would also work:
```php
return collect($this->resolveStack)->slice(0, -1)->contains(function ($resolving) {
    $object instanceof $resolving;
});
```

Reply:
It does, but it's a big performance hit: 

```
+-------------------------------------------------------+------------------+-------------------+
| subject                                               | tag:collect:mean | tag:original:mean |
+-------------------------------------------------------+------------------+-------------------+
| testResolvingCallbacksAreCalledForSpecificAbstracts   | 5,917.960μs      | 4,046.610μs       |
| testResolvingCallbacksAreCalledOnceForImplementations | 23,971.640μs     | 9,072.640μs       |
| testResolvingCallbacksAreCalledForNestedDependencies  | 99,831.230μs     | 61,386.800μs      |
+-------------------------------------------------------+------------------+-------------------+
```